src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -4872,12 +4873,16 @@ namespace Microsoft.FSharp.Core
             let PowGeneric (one, mul, value: 'T, exponent) = ComputePowerGenericInlined  one mul value exponent 
 
             let inline ComputeSlice bound start finish length =
-                match start, finish with
-                | None, None -> bound, bound + length - 1
-                | None, Some n when n >= bound  -> bound, n
-                | Some m, None when m <= bound + length -> m, bound + length - 1
-                | Some m, Some n -> m, n
-                | _ -> raise (System.IndexOutOfRangeException())
+                let low = 
+                    match start with
+                    | Some n when n >= bound -> n
+                    | _ -> bound
+                let high = 
+                    match finish with 
+                    | Some m when m < bound + length -> m
+                    | _ -> bound + length - 1
+
+                low, high
```

Comment:
Consider returning `struct(low, high)` (this could improve the current return type)

Reply:
Update: Seems like we can't use struct tuples in prim-types here. We investigated trying to optimize it by using outrefs, but poking around with the JIT assembly created by inlined functions reveals that the compiler/CLR is probably smart enough to not make the tuple, and hence returning a reference tuple is probably fine here.